[Merged by Bors] - feat(combinatorics/simple_graph/prod): add `locally_finite` instance for the box product of `locally_finite` graphs, plus two related lemmas

bottine opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

See discussion [here](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/.E2.9C.94.20Product.20of.20locally.20finite.20simple.20graphs)

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

CBirkbeck committed:
feat(number_theory/modular_forms/congruence_subgroups): Add definition of congruence subgroups. (#15159)

This contains the definition of a congruence subgroup of `SL(2,Z)` and defines Gamma1, Gamma0 and full level subgroups. It also contains some basic results about them.

mcdoll committed:
feat(analysis/schwartz): Definition of the Schwartz space (#15850)

This PR adds the definition of the Schwartz space and by abstract results also of tempered distributions. We prove basic algebraic and topological properties of the Schwartz space.



Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

digama0 committed:
fix(*): fix padding around notations (#16333)

I noticed a bunch of issues around padding in mathport, so I went through and fixed all the padding issues I could find in mathlib.

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

bottine committed:
The box product of two locally finite graphs is locally finite, and two related lemmas.

bottine committed:
layout

astrainfinita committed:
feat(algebra/big_operators/multiset): add `prod_map_le_prod_map` (#16418)

eric-wieser committed:
fix(data/nat/interval): do not dedup when implementing `finset.Icc` etc (#16423)

This means that `finset.Iic n` no longer has quadratic complexity.

`#eval (finset.Iic 200000).card` is now almost instant rather than taking a very long time.

astrainfinita committed:
chore(algebra/big_operators/basic): golf (#16427)

bottine committed:
work on pairs directly

khwilson committed:
feat(analysis/calculus/uniform_limits_deriv): Swap limits and derivatives (#14090)

This commit proves that the derivative of the pointwise limit of a series of functions is the limit of the derivatives when the derivatives converge uniformly in some closed ball.

This and #13500 are two fundamental theorems for bootstrapping the theory of Dirichlet series. This theory underlies my ongoing attempts to prove the density of squarefree integers.

kmill commented on src/combinatorics/simple_graph/prod.lean:
```diff
@@ -50,6 +50,17 @@ by rw [box_prod_adj, and_iff_left rfl, or_iff_left (λ h : H.adj b b ∧ _, h.1.
+             exists_eq_right_right],
+  simp_rw [and_comm _ (x.2 = b'), exists_eq_right_right],
+  finish,
+end
```
I suggest formulating this lemma as
```lean
lemma box_prod_neighbor_set (x : α × β) :
  (G □ H).neighbor_set x = (G.neighbor_set x.1) ×ˢ {x.2} ∪ {x.1} ×ˢ (H.neighbor_set x.2) :=
begin
  ext ⟨a', b'⟩,
  simp only [mem_neighbor_set, set.mem_union, box_prod_adj, set.mem_prod, set.mem_singleton_iff],
  simp only [eq_comm, and_comm],
end
```

kmill replied:
(I forgot to mention that one reason for the change in proof is that `finish` is not accepted in mathlib anymore since there are no plans at the moment to port it to Lean 4.)

kmill commented on src/combinatorics/simple_graph/prod.lean:
```diff
@@ -165,4 +176,35 @@ by { haveI := (nonempty_prod.1 h.nonempty).1, haveI := (nonempty_prod.1 h.nonemp
+               mem_neighbor_set,prod.mk.inj_iff,exists_eq_right_right] at q,
+    obtain ⟨⟨_,_,rfl,rfl⟩,⟨a,_,_,_⟩⟩ := q,
+    exact (a.ne (rfl)).elim, },
+end
```
With the suggested `neighbor_set` reformulation, there's a relatively straightforward proof of this lemma.

First, we should have the following lemma added somewhere, likely `data/set/finite.lean`:
```lean
lemma set.to_finset_prod (s : set α) (t : set β) [fintype s] [fintype t] :
  (s ×ˢ t).to_finset = s.to_finset ×ˢ t.to_finset :=
by { ext, simp }
```
Then, the proof of `box_prod_degree` can be simplified to
```lean
lemma box_prod_degree [decidable_eq α] [decidable_eq β]
  [locally_finite G] [locally_finite H] (x : α × β) :
  (G □ H).degree x = G.degree x.1 + H.degree x.2 :=
begin
  simp_rw [← card_neighbor_set_eq_degree, box_prod_neighbor_set,
    ← set.to_finset_card, set.to_finset_union],
  convert finset.card_disjoint_union _;
    simp only [set.to_finset_prod, finset.card_product, set.to_finset_card,
      set.card_singleton, mul_one, one_mul],
  { rintro ⟨_,_⟩ q,
    simp only [finset.inf_eq_inter, finset.mem_inter, finset.mem_product, set.mem_to_finset,
      mem_neighbor_set, set.mem_singleton_iff] at q,
    obtain ⟨⟨q, rfl⟩, ⟨rfl, _⟩⟩ := q,
    exact (q.ne rfl).elim, },
end
```

bottine replied:
I felt the `finset` lemma would be better inserted just after `theorem to_finset_compl` in `data/fintype.basic.lean`, but I think a finiteness instance is missing at this point… too bad.

bottine replied:
OK, I'm a bit at a loss for where to put this lemma. For now it's sort of floating in the `finset` section of `data/set/finite.lean`.
What I'd prefer is to move the `fintype` instance for `(s ×ˢ t)` in `data.fintype.basic.lean` and then this lemma there too, but even that is not trivial, because the natural place for the lemma (right after `to_finset_compl`) is actually **before** where the instance would rightfully go.

kmill replied:
Maybe the best place is right after the `fintype_prod` instance.

Given the definition for `fintype_prod`, I'm wondering whether `rfl` will actually prove this lemma -- could you check? (Edit: I checked, it doesn't. There seems to be a missing lemma around here...)

bottine replied:
OK, I moved it there.

kmill reviewed (commented):
Thanks for this PR! I have just a couple suggestions.

YaelDillies committed:
feat(combinatorics/simple_graph/triangle/basic): add ε-triangle-free far graph predicate (#12988)

Define the property of being triangle-free far. This comes up in the triangle counting and triangle removal lemmas.

Co-authored-by: Bhavik Mehta <bhavik.mehta8@gmail.com>



Co-authored-by: Kyle Miller <kmill31415@gmail.com>

Multramate committed:
feat(ring_theory/dedekind_domain/integer_unit): define S-integers and S-units (#15646)

jakelev committed:
feat(combinatorics/young_diagram): add transposes, rows and columns of Young diagrams (#16120)

Add transposes, rows (and row lengths), columns (and column lengths) of Young diagrams.

urkud committed:
refactor(data/{finite,fintype}): redefine `infinite` in terms of `finite`, move to `data.finite.defs` (#16390)

MichaelStollBayreuth committed:
feat(number_theory/legendre_symbol/jacobi_symbol): add new file with the definition and properties of the Jacobi symbol (#16395)

This PR consists of a new file, `number_theory.legendre_symbol.jacobi_symbol`, that contains a definition of the Jacobi symbol (and sets up notation `[a | b]ⱼ` for it) and then proves a number of results, e.g., multiplicativity in both arguments, expressions for the cases `a = -1`, `a = 2`, `a = -2`, and several versions of quadratic reciprocity. It also proves that the symbol depends on a only through its residue class `mod b` and that it depends on `b` only through its residue class `mod 4*a` (quadratic reciprocity is needed for the latter).

The code has already been quite thoroughly reviewed while it was part of #16290 (which then was reduced to the auxiliary results needed for the Jacobi symbol code).

Discussion on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Jacobi.20symbol/near/295816984)

YaelDillies committed:
feat(order/upper_lower): `Iic` is not bounded below (#16425)

Nonempty lower sets are not bounded below in a `no_min_order`, and dual.

astrainfinita committed:
feat(analysis/normed/group/basic): add `norm_multiset_sum_le` (#16419)

bottine committed:
Corrections after Kyle Miller

YaelDillies committed:
feat(analysis/{normed/group}/seminorm): Hom classes for seminorms (#16227)

Introduce `add_group_seminorm_class`, `group_seminorm_class`, `seminorm_class`, the hom classes of `add_group_seminorm`, `group_seminorm`, `seminorm`.

ADedecker committed:
feat(analysis/inner_product_space/basic): completion of an `inner_product_space` is an `inner_product_space` (#16407)

ADedecker committed:
feat(topology/algebra/module/weak_dual): `weak_space` is functorial (#16441)

mcdoll committed:
feat(linear_algebra/linear_pmap): add_action and mul_action (#16068)

This PR defines the instances `has_vadd`, `add_action` and `mul_action` for `linear_pmap`.



Co-authored-by: Moritz Doll <doll@uni-bremen.de>
Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

urkud committed:
feat(counterexamples/sorgenfrey_line): new file (#14820)

Define the Sorgenfrey line and prove that it is a normal topological space but its product with itself is not a normal topological space.

bottine committed:
move to_finset_prod just after fintype_prod

RemyDegenne committed:
feat(probability/stopping): generalize `mem_ℒp_stopped_value` (#16369)

Also rename many lemmas which had names `*_of_encodable` but took `countable` arguments since a recent PR which replaced all `encodable` by `countable` but did not touch the names.



Co-authored-by: Rémy Degenne <remydegenne@gmail.com>
Co-authored-by: RemyDegenne <remydegenne@gmail.com>

j-loreaux committed:
refactor(analysis/normed_space/spectrum): generalize Gelfand-Mazur away from `normed_division_ring` (#16448)

This generalizes the Gelfand-Mazur theorem so that it does not require the hypothesis that the algebra is a `normed_division_ring`, but instead is only a `normed_ring` in which the nonzero elements are precisely the units. This is important for applications of the Gelfand-Mazur theorem in which one may not be able to prove *a priori* that `∥a * b∥ = ∥a∥ * ∥b∥`, even though it may follow *a posteriori* from the Gelfand-Mazur isomorphism. An explicit example of this is the quotient of a complex Banach algebra by a maximal ideal, which is the use case that prompted this change.

YaelDillies committed:
feat(algebra/parity): Primes are not square (#16430)

and a few other basic results about `is_square`.

kmill commented on src/data/set/finite.lean:
```diff
@@ -310,6 +310,10 @@ instance fintype_prod (s : set α) (t : set β) [fintype s] [fintype t] :
 
+@[simp] lemma to_finset_prod (s : set α) (t : set β) [fintype s] [fintype t] :
+  (s ×ˢ t).to_finset = s.to_finset ×ˢ t.to_finset :=
+by { ext, simp only [set.mem_to_finset, set.mem_prod, finset.mem_product] }
```
I was thinking about this one more. I believe correct place to put it is where I put it in #16459. We can either wait for that to be merged (should be quick I hope) or you can copy what I have there and I'll delete that PR -- whatever you want to do!

bottine replied:
Let's wait for your PR then! But I'm confused:I believe I tried exactly that but got an error, which I understood to be because the `fintype` instance for the product is only defined at line 894. Wasn't that the case?

bottine replied:
Aha, nevermind, you added `[fintype (s ×ˢ t)]` as an assumption… neat!

kmill replied:
Yeah, it turns out these lemmas work better if everything that takes a `fintype` instance gets its own `fintype` argument. For example, with this lemma we want it to be able to apply to any `(s ×ˢ t).to_finset` term, even if its `fintype` instance isn't the standard one.

Also, #16459 has been merged!

bottine replied:
OK, I rebased and pushed. Maybe that was not the correct strategy, but anyway…

kmill replied:
It looks like this wasn't the correct strategy since GitHub thinks you've modified 143 files. I'll try to untangle what happened. (For mathlib I usually merge rather than rebase.)

kmill committed:
feat(data/fintype/basic): `fintype.to_finset_prod` (#16459)

urkud committed:
feat(topology/uniform_space/compact_separated): drop unneeded assumptions (#16457)

Lemmas in this file don't need `[separated_space _]` / `is_separated _` assumption. Also add `nhds_set` versions of some lemmas.

YaelDillies committed:
refactor(algebra/order/monoid): Remove redundant field from `ordered_cancel_comm_monoid` (#16445)

`mul_left_cancel : ∀ a b c, a + b = a + c → b = c` can be inferred from `le_of_mul_le_mul_left : ∀ a b c, a + b ≤ a + c → b ≤ c` and antisymmetry.

RemyDegenne committed:
chore(probability/martingale): move the optional stopping theorem to its own file (#16464)

The idea is to reserve martingale/basic.lean to really basic API about (sub/super)martingales.

vihdzp committed:
feat(data/{seq,stream}): cons_injective2 (#15832)

Also some simple API lemmas for `nth`
I didn't change the style of the `stream/init` file.





Co-authored-by: Yakov Pechersky <ypechersky@treeline.bio>

RemyDegenne committed:
feat(measure_theory/function/conditional_expectation): conditional expectation with respect to the bot sigma algebra (#16342)

Prove that for a probability measure, `μ[f|⊥] = λ _, ∫ x, f x ∂μ`.



Co-authored-by: RemyDegenne <remydegenne@gmail.com>

erdOne committed:
feat(linear_algebra/finsupp): A variant of `finsupp.total` for `fintype`s. (#15462)




Co-authored-by: Andrew Yang <36414270+erdOne@users.noreply.github.com>

vihdzp committed:
refactor(set_theory/basic): match `x < ℵ₀` lemmas with `x ≤ ℵ₀` lemmas (#15662)

* Mark `cardinal.lt_aleph_0_iff_set_finite` as `@[simp]` lemma.
* Add `cardinal.lt_aleph_0_iff_subtype_finite` and `cardinal.mk_le_aleph_0_iff`; drop `cardinal.encodable_iff`.
* Rename `cardinal.mk_set_le_aleph_0` to `cardinal.le_aleph_0_iff_set_countable`.
* Rename `cardinal.mk_subtype_le_aleph_0` to `cardinal.le_aleph_0_iff_subtype_countable`.
* Make `first_order.language.countable` protected.
* Use `[countable _]` instead of `[encodable _]` or `[nonempty (encodable _)]` here and there.



Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

linesthatinterlace committed:
feat(data/real/ennreal): Add `to_real_{min, sup, inf}` (#16233)

This adds `to_real_min` to match `to_real_max`, and adds `inf` and `sup` spellings of these lemmas.


Co-authored-by: Wrenna Robson <34025592+linesthatinterlace@users.noreply.github.com>

YaelDillies committed:
feat(analysis/normed/group/seminorm): Group norms (#16237)

Define (additive) group norms, which are group seminorms `f` such that `f x = 0 → x = 0` (resp. `f x = 0 → x = 1`), along with their hom classes.

astrainfinita committed:
feat(data/polynomial/algebra_map): add `aeval_X_left_apply` (#16433)

eric-wieser committed:
chore(data/real/cau_seq_completion): golf using existing lemmas (#16469)

This extracts `cau_seq.mul_equiv_mul` and `cau_seq.sub_equiv_sub` into standalone lemmas, and uses the existing lemmas for `add` and `neg` rather than reproving the result from scratch.

Paul-Lez committed:
feat(algebra/algebra/basic): add two simp lemmas (#16096)

Two simp lemmas that are useful for dealing with the group of automorphisms of an algebra.
I also had to fix two proofs from `field_theory/krull_topology`.

hrmacbeth committed:
feat(analysis/inner_product_space/pi_L2): change of basis matrix between orthonormal bases is unitary (#16474)

Formalized as part of the Sphere Eversion project.

PatrickMassot committed:
chore(linear_algebra): rename dual_pair (#16482)

This only renames `dual_pair` to `module.dual_bases`, avoiding to put a very generic name into the root name space. This `dual_pair` was used only in the archive, I introduced it a very long time ago to streamline the proof of the sensitivity conjecture.

pimotte committed:
feat(data/nat/{choose/multinomial,factorial/big_operations}): add multinomial coefficient definition with basic lemma's (#16170)

Chose a definition based on finset and a function for multiplicity. This allows the definition to be used in various context, including in a multiset context, where multiset.count yields the function describing multiplicity. Supporting lemma's concerning factorials are introduced in a new file, since importing big_operators in the main factorial file yields a circular import.

Co-authored-by: Kyle Miller <kmill31415@gmail.com>

Paul-Lez committed:
feat(ring_theory/dedekind_domain/ideal): construct map between the sets of prime factors of ideal `I` and  `J` induced by an isomorphism between `R/I` and `S/J` (#16455)

These lemmas generalize results that were originally in #15000.

Co-authored-by: Anne Baanen  <vierkantor@vierkantor.com>

j-loreaux committed:
feat(data/complex/module): define `real_part` and `imaginary_part` (#16438)

In a generic `star_module` one always has a decomposition of any element into a `self_adjoint_part` and a `skew_self_adjoint` part, but in a star module over `ℂ` we can instead decompose any element into a `real_part` and an `imaginary_part`, both of which are self-adjoint. Here we define these as `ℝ`-linear maps from the star module into the type of its self-adjoint elements and describe the basic relationships between these maps. The decomposition into real and imaginary parts is often useful for reducing arguments about elements of a star module over `ℂ` to the case when the element is self-adjoint.

bottine committed:
remove fintype_prod

kmill committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into bottine/box_prod_locally_finite

kmill committed:
fully generalize fintype instances

kmill committed:
decidability linter
